### PR TITLE
Methods are documented on class pages

### PR DESCRIPTION
**Pull request by Eric-Arellano**
Currently, methods have their own HTML page. This PR changes them to live on the class page.

Motivations:

1. Speed up our docs build. Before: 8 minutes, 20 seconds. After: 4 minutes, 30 seconds.
2. Better user experience. While it's true we sometimes have very long method entries, most of our methods are small and it's a frustrating user experience to have to change pages so much. Compare this to e.g. Python docs, where the entire module is documented on a single page. An IBM design team suggested we make this change.

A follow-up will move functions to their module page, rather than dedicated pages.

### No redirects

We decided it's not worth adding redirects for method pages. Sphinx will already properly update all cross-references to the right place, and we don't expect there to be many external links to method pages.

### CODE REVIEW

**Review by jakelishman — COMMENTED** [review]
This looks pretty inline with the minimal change I was expecting.

Is it worth completely removing the loops from the `class` summary templates, and switching the `:no-members:` option to `:members:`, so `autodoc` manages it all itself?  I've found raw `autodoc` to be quite sane, and for it to generally improve over time as well.

**Comment by coruscating** [comment]
Is it possible to add links on the right sidebar for each section (Attributes, Methods, etc.)? For example, `QuantumCircuit` is now 40+ pages long without any subheadings on the right sidebar, and I find it pretty difficult to navigate.

**Comment by jakelishman** [comment]
For `QuantumCircuit` in particular, we'll probably want to write its documentation a bit more manually than the autosummary format because it's so long, it could really do with a bit more human interaction to group the methods into logical groups.

Changing the `.. rubric::` lines to a heading directive instead (e.g. underlining "Attributes" with `------------`) should add toctree entries to the page, though I can't remember off the top of my head how well `docutils` will interact with the text-templating that `autosummary` does to insert these if there's already headings in the summarised page.  I _think_ it works ok because `autosummary` always creates a new page, but it'd need testing.

**Comment by Eric-Arellano** [comment]
> Is it worth completely removing the loops from the class summary templates, and switching the :no-members: option to :members:, so autodoc manages it all itself?

I don't think so because we want the `.. rubric` directive to add the heading for methods vs attributes.

--

@coruscating that's a Pytorch problem. The example docs use the exact same templates at this PR:

Screenshot 2023-07-20 at 9 59 47 AM

Screenshot 2023-07-20 at 10 00 44 AM

This PR and the followup to remove functions both unblock us switching to Furo in the metapackage - we'll switch over right away after. 

Further, this change doesn't go "live" until we do the final 0.25 release because the metapackage builds off of releases, not `main`. So, I don't think this UX problem will actually impact users in reality.

**Comment by coruscating** [comment]
@Eric-Arellano That's good to see the new theme automatically adds links to individual attributes and methods, though I'd prefer to have actual headings like "Methods" and "Attributes" that they're listed under, otherwise it'll be a very long unsorted list on some pages.

@jakelishman We tried changing `.. rubric::` to headings in Experiments, but that broke things because of unhappy interactions between napoleon and Sphinx: https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1077#discussion_r1137831502. I'm not sure if this would also be a problem here.

**Comment by jakelishman** [comment]
Helena: the reason it might be different for you compared to us is that you guys were (or at least appeared to be) modifying a docstring that was embedded into more free-form other pages, where you didn't have control on what other section-heading constructs were used.  In the current setup, `autosummary` generates a new page for anything that triggers the `class.rst` template, and that page has a _very_ fixed structure, so there (probably) aren't any other headings of the same logic level with which we might conflict if we unilaterally pick a punctuation mark to use in the heading here.

**Comment by jakelishman** [comment]
Eric: in my vein of "raw `autodoc` is usually pretty sensible": [there's an option to control the grouping](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_member_order) (though admittedly, I don't know if it adds section headings).

**Comment by Eric-Arellano** [comment]
Good find @jakelishman. But looks like we don't want to use it. With this template:

```rst
{{ objname | escape | underline }}

.. currentmodule:: {{ module }}

.. autoclass:: {{ objname }}
   :members:
   :show-inheritance:
```

That means we won't have rubrics/headers and the members will appear as subheaders rather than top-level entries:

Screenshot 2023-07-20 at 11 12 26 AM

--

> though I'd prefer to have actual headings like "Methods" and "Attributes" that they're listed under, otherwise it'll be a very long unsorted list on some pages.

I was able to get headers working rather than rubrics:

```rst
{% block attributes_summary %}
  {% if attributes %}
   Attributes
   ----------
    {% for item in attributes %}
   .. autoattribute:: {{ item }}
    {%- endfor %}
  {% endif %}
{% endblock -%}
```

But, it doesn't impact the right sidebar at all. 

Screenshot 2023-07-20 at 11 15 54 AM

**Comment by jakelishman** [comment]
Ah, that's a shame.  I wonder if `autodoc` would consider that as a feature request, if it's the only thing keeping us from using it.  They've generally been pretty receptive to me in the past, I think - if we like the idea, I can take it up with them.

**Comment by Eric-Arellano** [comment]
> if we like the idea, I can take it up with them.

As long as we, of course, don't block this PR on it, sure! Upstreaming config is always nice.

**Comment by jakelishman** [comment]
Yeah, totally fine by me - I don't even think we're using Sphinx 7 yet.

**Comment by jakelishman** [comment]
I'd have to look into why changing the templates to insert a heading doesn't impact the sidebar, but whether or not the documented objects themselves get inserted into the toctree is controlled by the [`toc_object_entries` global Sphinx option](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-toc_object_entries).

**Comment by Eric-Arellano** [comment]
> but whether or not the documented objects themselves get inserted into the toctree is controlled by the [toc_object_entries global Sphinx option](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-toc_object_entries).

Yeah, and we're using the default of True already.

--

To clarify, is anything blocking this PR from approval? (Other than time / trying to get out rc1, of course)

**Review by jakelishman — APPROVED** [review]
Not from my side - I'd forgotten I hadn't done it.

**Comment by jakelishman** [comment]
I'm taking this out the merge queue because it'd use time that we need for the rc1 release right now.  We can add it after that.

**Pull request merged**
